source/jormungandr/jormungandr/scenarios/experimental.py

```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
+                g.destinations_places_crowfly.update(future.get())
+
+            orig_futures, dest_futures = worker.get_routing_matrix_futures(g.requested_origin,
+                                                                           g.requested_destination,
+                                                                           g.origins_places_crowfly,
+                                                                           g.destinations_places_crowfly)
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
+                g.origins_fallback.update(future.get())
+            for future in gevent.iwait(dest_futures):
+                g.destinations_fallback.update(future.get())
+
+            orig_futures, dest_futures = worker.get_update_crowfly_duration_futures()
+            gevent.joinall(orig_futures + dest_futures)
+            def _updater(_futures, fb, crow_fly_stop_points):
+                for f in gevent.iwait(_futures):
+                    resp = f.get()
+                    crow_fly_stop_points |= resp[0]
+                    for mode in (mode for mode in resp[1] if mode in fb):
+                        fb[mode].update(resp[1][mode])
+
+            _updater(orig_futures, g.origins_fallback, crow_fly_stop_points)
+            _updater(dest_futures, g.destinations_fallback, crow_fly_stop_points)
+
+
         resp = []
         journey_parameters = create_parameters(request)
-        for dep_mode, arr_mode in krakens_call:
-            #todo: this is probably shared between multiple thread
-            self.nb_kraken_calls += 1
-            direct_path = self._get_direct_path(instance,
-                                                dep_mode,
-                                                g.requested_origin,
-                                                g.requested_destination,
-                                                request['datetime'],
-                                                request['clockwise'],
-                                                request)
-            if direct_path.journeys:
-                journey_parameters.direct_path_duration = direct_path.journeys[0].durations.total
-                #Since _get_direct_method returns a reference instead of an object we don't add in response
-                #if already exists.
-                if direct_path not in resp:
-                    direct_path.journeys[0].internal_id = "dp_{resp}-{j}".format(resp=self.nb_kraken_calls, j=0)
-                    resp.append(direct_path)
-            else:
-                journey_parameters.direct_path_duration = None
 
-            fall_back_origins = g.origins_fallback.get(dep_mode)
-            fall_back_destinations = g.destinations_fallback.get(arr_mode)
-            if not fall_back_origins or not fall_back_destinations or not request.get('max_duration', 0):
+        futures = worker.get_direct_path_futures(self._get_direct_path,
```

Comment:
Comment to add:
we compute the direct path for this journey, their time will be used to initialized our PT calls